c/tskit/trees.c

```diff
@@ -8145,3 +8165,339 @@ tsk_treeseq_extend_edges(const tsk_treeseq_t *self, int max_iter,
+    nodes_parent = tsk_malloc(num_nodes * sizeof(*nodes_parent));
+    nodes_sample = tsk_malloc(num_nodes * (tsk_size_t) inp * sizeof(*nodes_sample));
+    sample_count = tsk_malloc(num_nodes * (tsk_size_t) inp * sizeof(*sample_count));
+    if (nodes_parent == NULL || nodes_sample == NULL || sample_count == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    tsk_memset(nodes_sample, 0, num_nodes * (tsk_size_t) inp * sizeof(*nodes_sample));
+    for (i = 0; i != (tsk_id_t) num_nodes; i++) {
+        if (nodes_sample_set[i] != TSK_NULL) {
+            nodes_sample[i * inp + nodes_sample_set[i]] = 1;
+        }
+    }
+    tsk_memcpy(sample_count, nodes_sample,
+        num_nodes * (tsk_size_t) inp * sizeof(*sample_count));
+    for (i = 0; i != (tsk_id_t) num_nodes; i++) {
+        nodes_parent[i] = TSK_NULL;
+    }
+
+    /* map nodes to time windows */
+    nodes_time_window = tsk_malloc(num_nodes * sizeof(*nodes_time_window));
+    if (nodes_time_window == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    ret = get_time_windows_index_map(
+        self, &num_time_windows, time_windows, nodes_time_window);
+    if (ret != 0) {
+        goto out;
+    }
+
+    /* initialize outputs */
+    row = (tsk_id_t) num_time_windows;
+    inside = tsk_malloc((tsk_size_t) col * sizeof(inside));
+    outside = tsk_malloc((tsk_size_t) col * sizeof(outside));
+    if (inside == NULL || outside == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    coalescing_pairs
+        = tsk_malloc((tsk_size_t) row * (tsk_size_t) col * sizeof(coalescing_pairs));
+    nodes_weight
+        = tsk_malloc((tsk_size_t) row * (tsk_size_t) col * sizeof(nodes_weight));
+    if (coalescing_pairs == NULL || nodes_weight == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    tsk_memset(coalescing_pairs, 0,
+        (tsk_size_t) row * (tsk_size_t) col * sizeof(*coalescing_pairs));
+
+    tsk_memset(&tree_pos, 0, sizeof(tree_pos));
+    ret = tsk_tree_position_init(&tree_pos, self, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    w = 0;
+    while (true) {
+        tsk_tree_position_next(&tree_pos);
+        if (tree_pos.index == TSK_NULL) {
+            break;
+        }
+
+        left = tree_pos.interval.left;
+        right = tree_pos.interval.right;
+        remaining_span = sequence_length - left;
+
+        for (u = tree_pos.out.start; u != tree_pos.out.stop; u++) {
+            e = tree_pos.out.order[u];
+            p = tables->edges.parent[e];
+            c = tables->edges.child[e];
+            nodes_parent[c] = TSK_NULL;
+            for (i = 0; i != inp; i++) { /* samples beneath child */
+                inside[i] = sample_count[c * inp + i];
+            }
+            while (p != TSK_NULL) {
+                v = nodes_time_window[p];
+                if (v != TSK_NULL) {
+                    for (i = 0; i != inp; i++) { /* samples beneath sibs */
```

Comment:
I would rather use ``num_sample_sets`` here (or whatever it is) with a cast if necessary. It's easier to read, and there's no run-time cost to having a cast. The casts are really just there to keep the compiler happy.